File: encre/display/GDEP073E01.py
# Modified from:
# https://github.com/pimoroni/inky/blob/main/library/inky/inky_ac073tc1a.py

# MIT License
#
# Copyright (c) 2018 Pimoroni Ltd.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import time
import warnings
import numpy
import numpy.typing

import RPi.GPIO as GPIO
import spidev

from display_protocol import Display


class GDEP073E01:
    DC_PIN = 22
    RESET_PIN = 27
    BUSY_PIN = 17

    BLACK = 0
    WHITE = 1
    YELLOW = 2
    RED = 3
    # Actually 4 is unused
    BLUE = 4 # Actually 5
    GREEN = 5 # Actually 6

    WIDTH = 800
    HEIGHT = 480

    _PSR = 0x00
    _PWR = 0x01
    _POF = 0x02
    _POFS = 0x03
    _PON = 0x04
    _BTST1 = 0x05
    _BTST2 = 0x06
    _BTST3 = 0x08
    _DTM = 0x10
    _DRF = 0x12
    _IPC = 0x13
    _PLL = 0x30
    _TSE = 0x41
    _CDI = 0x50
    _TCON = 0x60
    _TRES = 0x61
    _VDCS = 0x82
    _T_VDCS = 0x84
    _AGID = 0x86
    _CMDH = 0xAA
    _CCSET = 0xE0
    _PWS = 0xE3
    _TSSET = 0xE6

    _SPI_COMMAND = 0
    _SPI_DATA = 1

    buf: numpy.typing.NDArray[numpy.uint8]

    def __init__(self, *, cs_channel=0, dc_pin=DC_PIN, reset_pin=RESET_PIN, busy_pin=BUSY_PIN):
        """
        :param cs_channel: channel for SPI communication
        :param dc_pin: data/command pin
        :param reset_pin: device reset pin
        :param busy_pin: device busy/wait pin
        """
        self._spi_bus = None
        self._gpio = None

        self.buf = numpy.zeros((GDEP073E01.HEIGHT, GDEP073E01.WIDTH), dtype=numpy.uint8)

        self.dc_pin = dc_pin
        self.reset_pin = reset_pin
        self.busy_pin = busy_pin
        self.cs_channel = cs_channel

    def show(self):
        """
        Show buffer on display.

        :param busy_wait: If True, wait for display update to finish before returning.
        """
        buf = self.buf.flatten()

        # Remove invalid colors, correct for unused index 4
        buf = numpy.clip(buf, 0, 5)
        buf[buf >= 4] += 1

        buf = ((buf[::2] << 4) & 0xF0) | (buf[1::2] & 0x0F)

        self._update(buf.astype('uint8').tolist())

    def _setup(self):
        """Set up IO if needed and reset display."""

        if self._gpio is None:
            self._gpio = GPIO
            self._gpio.setmode(self._gpio.BCM)
            self._gpio.setwarnings(False)
            self._gpio.setup(self.dc_pin, self._gpio.OUT, initial=self._gpio.LOW, pull_up_down=self._gpio.PUD_OFF)
            self._gpio.setup(self.reset_pin, self._gpio.OUT, initial=self._gpio.HIGH, pull_up_down=self._gpio.PUD_OFF)
            self._gpio.setup(self.busy_pin, self._gpio.IN, pull_up_down=self._gpio.PUD_OFF)

        if self._spi_bus is None:
            self._spi_bus = spidev.SpiDev()
            self._spi_bus.open(0, self.cs_channel)
            self._spi_bus.max_speed_hz = 5000000

        self._gpio.output(self.reset_pin, self._gpio.LOW)
        time.sleep(0.1)
        self._gpio.output(self.reset_pin, self._gpio.HIGH)
        time.sleep(0.1)

        self._gpio.output(self.reset_pin, self._gpio.LOW)
        time.sleep(0.1)
        self._gpio.output(self.reset_pin, self._gpio.HIGH)

        self._busy_wait(1.0)

        # Sending init commands to display
        self._send_command(GDEP073E01._CMDH, [0x49, 0x55, 0x20, 0x08, 0x09, 0x18])
        self._send_command(GDEP073E01._PWR, [0x3F, 0x00, 0x32, 0x2A, 0x0E, 0x2A])
        self._send_command(GDEP073E01._PSR, [0x5F, 0x69])
        self._send_command(GDEP073E01._POFS, [0x00, 0x54, 0x00, 0x44])
        self._send_command(GDEP073E01._BTST1, [0x40, 0x1F, 0x1F, 0x2C])
        self._send_command(GDEP073E01._BTST2, [0x6F, 0x1F, 0x16, 0x25])
        self._send_command(GDEP073E01._BTST3, [0x6F, 0x1F, 0x1F, 0x22])
        self._send_command(GDEP073E01._IPC, [0x00, 0x04])
        self._send_command(GDEP073E01._PLL, 0x02)
        self._send_command(GDEP073E01._TSE, 0x00)
        self._send_command(GDEP073E01._CDI, 0x3F)
        self._send_command(GDEP073E01._TCON, [0x02, 0x00])
        self._send_command(GDEP073E01._TRES, [0x03, 0x20, 0x01, 0xE0])
        self._send_command(GDEP073E01._VDCS, 0x1E)
        self._send_command(GDEP073E01._T_VDCS, 0x01)
        self._send_command(GDEP073E01._AGID, 0x00)
        self._send_command(GDEP073E01._PWS, 0x2F)
        self._send_command(GDEP073E01._CCSET, 0x00)
        self._send_command(GDEP073E01._TSSET, 0x00)

    def _busy_wait(self, timeout=40):
        # If the busy_pin is *high* (pulled up by host)
        # then assume we're not getting a signal from display
        # and wait the timeout period to be safe.
        if self._gpio.input(self.busy_pin):
            warnings.warn("Busy Wait: Held high. Waiting for {:0.2f}s".format(timeout))
            time.sleep(timeout)
            return

        # If the busy_pin is *low* (pulled down by display)
        # then wait for it to high.
        t_start = time.time()
        while not self._gpio.input(self.busy_pin):
            time.sleep(0.01)
            if time.time() - t_start >= timeout:
                warnings.warn("Busy Wait: Timed out after {:0.2f}s".format(time.time() - t_start))
                return

    def _update(self, buf):
        self._setup()

        self._send_command(GDEP073E01._DTM, buf)

        self._send_command(GDEP073E01._PON)
        self._busy_wait(0.4)

        self._send_command(GDEP073E01._DRF, 0x00)
        self._busy_wait(45.0)  # 41 seconds in testing

        self._send_command(GDEP073E01._POF, 0x00)
        self._busy_wait(0.4)

    def _spi_write(self, dc, values):
        self._gpio.output(self.dc_pin, dc)

        if type(values) is str:
            values = [ord(c) for c in values]

        for byte_value in values:
            self._spi_bus.xfer([byte_value])

    def _send_command(self, command, data=None):
        self._spi_write(GDEP073E01._SPI_COMMAND, [command])
        if data is not None:
            self._send_data(data)

    def _send_data(self, data):
        try:
            _ = iter(data)
        except TypeError:
            data = [data]
        self._spi_write(GDEP073E01._SPI_DATA, data)


def init_display(_) -> Display:
    return GDEP073E01()

File: microfiche/src/GDEP073E01.c
#include "GDEP073E01.h"

#include <hardware/gpio.h>
#include <hardware/spi.h>
#include <pico/time.h>

#define DC_PIN 22
#define RESET_PIN 21
#define BUSY_PIN 20

#define BLACK 0
#define WHITE 1
#define YELLOW 2
#define RED 3
// Actually 4 is unused
#define BLUE 4 // Actually 5
#define GREEN 5 // Actually 6

#define CMD_PSR 0x00
#define CMD_PWR 0x01
#define CMD_POF 0x02
#define CMD_POFS 0x03
#define CMD_PON 0x04
#define CMD_BTST1 0x05
#define CMD_BTST2 0x06
#define CMD_DSLP 0x07
#define CMD_BTST3 0x08
#define CMD_DTM 0x10
#define CMD_DRF 0x12
#define CMD_IPC 0x13
#define CMD_PLL 0x30
#define CMD_TSE 0x41
#define CMD_CDI 0x50
#define CMD_TCON 0x60
#define CMD_TRES 0x61
#define CMD_VDCS 0x82
#define CMD_T_VDCS 0x84
#define CMD_AGID 0x86
#define CMD_CMDH 0xAA
#define CMD_CCSET 0xE0
#define CMD_PWS 0xE3
#define CMD_TSSET 0xE6

#define SPI_COMMAND false
#define SPI_DATA true

static void setup_pin(uint pin, enum gpio_dir direction, bool initial) {
    gpio_init(pin);
    gpio_set_dir(pin, direction);
    if (direction == GPIO_OUT) {
        gpio_put(pin, initial);
    }
}

static void busy_wait(uint32_t timeout_ms) {
    if (gpio_get(BUSY_PIN)) {
        sleep_ms(timeout_ms);
        return;
    }

    const uint64_t timeout_us = (uint64_t)timeout_ms * 1000;
    const uint64_t start_time = time_us_64();
    do {
        sleep_ms(10);
        if (time_us_64() - start_time >= timeout_us) {
            return;
        }
    } while (!gpio_get(BUSY_PIN));
}

static void write_to_spi(bool dc, const uint8_t *data, size_t length) {
    gpio_put(DC_PIN, dc);

    spi_write_blocking(spi0, data, length);
}

static void send_command(uint8_t command) {
    write_to_spi(SPI_COMMAND, &command, sizeof(command));
}

static void send_data(const uint8_t *data, size_t length) {
    write_to_spi(SPI_DATA, data, length);
}

#define send_command_with_data(command, ...) \
    do { \
        send_command(command); \
        const uint8_t cmd_data[] = {__VA_ARGS__}; \
        send_data(cmd_data, sizeof(cmd_data)); \
    } while (false)

static void setup() {
    static bool initialized = false;
    if (!initialized) {
        setup_pin(DC_PIN, GPIO_OUT, false);
        setup_pin(RESET_PIN, GPIO_OUT, true);
        setup_pin(BUSY_PIN, GPIO_IN, false);

        gpio_set_function(19, GPIO_FUNC_SPI);
        gpio_set_function(18, GPIO_FUNC_SPI);
        gpio_set_function(17, GPIO_FUNC_SPI);
        gpio_set_function(16, GPIO_FUNC_SPI);
        spi_init(spi0, 5000000);

        initialized = true;
    }

    gpio_put(RESET_PIN, false);
    sleep_ms(100);
    gpio_put(RESET_PIN, true);
    sleep_ms(100);
    gpio_put(RESET_PIN, false);
    sleep_ms(100);
    gpio_put(RESET_PIN, true);

    busy_wait(1000);

    send_command_with_data(CMD_CMDH, 0x49, 0x55, 0x20, 0x08, 0x09, 0x18);
    send_command_with_data(CMD_PWR, 0x3F, 0x00, 0x32, 0x2A, 0x0E, 0x2A);
    send_command_with_data(CMD_PSR, 0x5F, 0x69);
    send_command_with_data(CMD_POFS, 0x00, 0x54, 0x00, 0x44);
    send_command_with_data(CMD_BTST1, 0x40, 0x1F, 0x1F, 0x2C);
    send_command_with_data(CMD_BTST2, 0x6F, 0x1F, 0x16, 0x25);
    send_command_with_data(CMD_BTST3, 0x6F, 0x1F, 0x1F, 0x22);
    send_command_with_data(CMD_IPC, 0x00, 0x04);
    send_command_with_data(CMD_PLL, 0x02);
    send_command_with_data(CMD_TSE, 0x00);
    send_command_with_data(CMD_CDI, 0x3F);
    send_command_with_data(CMD_TCON, 0x02, 0x00);
    send_command_with_data(CMD_TRES, 0x03, 0x20, 0x01, 0xE0);
    send_command_with_data(CMD_VDCS, 0x1E);
    send_command_with_data(CMD_T_VDCS, 0x01);
    send_command_with_data(CMD_AGID, 0x00);
    send_command_with_data(CMD_PWS, 0x2F);
    send_command_with_data(CMD_CCSET, 0x00);
    send_command_with_data(CMD_TSSET, 0x00);
}

void GDEP073E01_write_image(const uint8_t *color_bits) {
    setup();

    send_command(CMD_DTM);

    const uint8_t color_mask = (1 << GDEP073E01_BITS_PER_COLOR) - 1;
    uint8_t color_pair = 0;
    for (uint32_t i = 0, bit_index = 0; i < GDEP073E01_WIDTH * GDEP073E01_HEIGHT;
            i++, bit_index += GDEP073E01_BITS_PER_COLOR) {
        const uint32_t byte_index = bit_index >> 3;
        const uint32_t bit_offset = bit_index & 7;
        uint32_t color = (color_bits[byte_index] >> bit_offset) & color_mask;

        const uint32_t next_byte_index = (bit_index + GDEP073E01_BITS_PER_COLOR) >> 3;
        if (byte_index != next_byte_index) {
            const uint32_t next_bit_offset = 8 - bit_offset;
            const uint32_t next_mask = color_mask >> next_bit_offset << next_bit_offset;
            color |= (color_bits[next_byte_index] << next_bit_offset) & next_mask;
        }

        color = MAX(MIN(color, 5), 0);
        if (color >= 4) {
            color++;
        }

        color_pair = (color_pair << 4) | color;

        if (i & 1) {
            send_data(&color_pair, sizeof(color_pair));
        }
    }

    send_command(CMD_PON);
    busy_wait(400);

    send_command_with_data(CMD_DRF, 0x00);
    busy_wait(45000);

    send_command_with_data(CMD_POF, 0x00);
    busy_wait(400);

    send_command_with_data(CMD_DSLP, 0xA5);
}
